activestorage/app/controllers/active_storage/representations/base_controller.rb

```diff
@@ -6,6 +6,18 @@ class ActiveStorage::Representations::BaseController < ActiveStorage::BaseContro
   before_action :set_representation
 
   private
+    def set_blob
+      scope = ActiveStorage::Blob
+
+      if ActiveStorage.track_variants && ActiveStorage::Blob.strict_loading_by_default?
+        scope = scope.includes(variant_records: { image_attachment: :blob }, preview_image_attachment: :blob)
```

Comment:
I haven't tested it, but would `scope.with_all_variant_records` work? If it does then you could potentially remove the `if` statement.

Reply:
https://github.com/rails/rails/pull/40842 added ^ scope

Reply:
I was thinking about configurations where the Variant is not VariantRecord to not introduce extra queries.

Reply:
Oh, i see what you talking about. Will test

Reply:
I think this will not work for the previews controller

Reply:
Rather than stuff all this logic into this controller we should add a scope method to `ActiveStorage::SetBlob`, e.g:

``` ruby
# frozen_string_literal: true

module ActiveStorage::SetBlob #:nodoc:
  extend ActiveSupport::Concern

  included do
    before_action :set_blob
  end

  private
    def set_blob
      @blob = blob_scope.find_signed!(params[:signed_blob_id] || params[:signed_id])
    rescue ActiveSupport::MessageVerifier::InvalidSignature
      head :not_found
    end

    def blob_scope
      ActiveStorage::Blob
    end
end
```

This would allow us to move the logic into `ActiveStorage::Blob` as a class method:

``` ruby
class ActiveStorage::Blob < ActiveRecord::Record
  class << self
    def scope_for_strict_loading
      if strict_loading_by_default? && ActiveStorage.track_variants
        includes(variant_records: { image_attachment: :blob }, preview_image_attachment: :blob)
      else
        all
      end
    end
  end
end
```

Then the implementation in this controller is just overriding the `blob_scope` method:

``` ruby
def blob_scope
  ActiveStorage::Blob.scope_for_strict_loading
end
```

I've not double checked this but it should all be fine assuming the `includes` is correct.